localstack/services/s3/models_native.py

```diff
@@ -0,0 +1,486 @@
+    MultipartUploadId,
+    NoSuchKey,
+    NoSuchVersion,
+    NotificationConfiguration,
+    ObjectKey,
+    ObjectLockConfiguration,
+    ObjectLockLegalHoldStatus,
+    ObjectLockMode,
+    ObjectOwnership,
+    ObjectStorageClass,
+    ObjectVersionId,
+    Owner,
+    Payer,
+    Policy,
+    PublicAccessBlockConfiguration,
+    ReplicationConfiguration,
+    Restore,
+    ServerSideEncryption,
+    ServerSideEncryptionRule,
+    Size,
+    SSEKMSKeyId,
+    StorageClass,
+    WebsiteConfiguration,
+    WebsiteRedirectLocation,
+)
+from localstack.services.s3.storage import LockedSpooledTemporaryFile
+from localstack.services.s3.utils import (
+    ParsedRange,
+    get_owner_for_account_id,
+    iso_8601_datetime_without_milliseconds_s3,
+    rfc_1123_datetime,
+)
+from localstack.services.stores import (
+    AccountRegionBundle,
+    BaseStore,
+    CrossAccountAttribute,
+    CrossRegionAttribute,
+)
+
+# TODO: beware of timestamp data, we need the snapshot to be more precise for S3, with the different types
+# moto had a lot of issue with it? not sure about our parser/serializer
+
+LOG = logging.getLogger(__name__)
+
+
+# note: not really a need to use a dataclass here, as it has a lot of fields, but only a few are set at creation
+class S3Bucket:
+    name: BucketName
+    bucket_account_id: AccountId
+    bucket_region: BucketRegion
+    creation_date: datetime
+    multiparts: dict[MultipartUploadId, Any]  # TODO
+    objects: Union["KeyStore", "VersionedKeyStore"]
+    versioning_status: Literal[None, "Enabled", "Disabled"]
+    lifecycle_rules: LifecycleRules
+    policy: Optional[Policy]
+    website_configuration: WebsiteConfiguration
+    acl: str  # TODO: change this
+    cors_rules: CORSConfiguration
+    logging: LoggingEnabled
+    notification_configuration: NotificationConfiguration
+    payer: Payer
+    encryption_rule: Optional[
+        ServerSideEncryptionRule
+    ]  # TODO validate if there can be more than one rule
+    public_access_block: PublicAccessBlockConfiguration
+    accelerate_status: BucketAccelerateStatus
+    object_ownership: ObjectOwnership
+    object_lock_configuration: Optional[ObjectLockConfiguration]
+    object_lock_enabled: bool
+    intelligent_tiering_configuration: dict[IntelligentTieringId, IntelligentTieringConfiguration]
+    analytics_configuration: dict[AnalyticsId, AnalyticsConfiguration]
+    replication: ReplicationConfiguration
+    owner: Owner
+
+    # set all buckets parameters here
+    def __init__(
+        self,
+        name: BucketName,
+        account_id: AccountId,
+        bucket_region: BucketRegion,
+        acl=None,  # TODO: validate ACL first, create utils for validating and consolidating
+        object_ownership: ObjectOwnership = None,
+        object_lock_enabled_for_bucket: bool = None,
+    ):
+        self.name = name
+        self.bucket_account_id = account_id
+        self.bucket_region = bucket_region
+        self.objects = KeyStore()
```

Comment:
Q: I think haven't fully understood the `KeyStore`/`StorageBackend` separation and whether it is needed. I suppose one is for managing the metadata, and one for the actual object data?

Reply:
Basically, the basic `KeyStore` could only be a `dict`, but it's there to hide away to complexity of the `VersionedKeyStore`. However, this can't really be mixed with the `StorageBackend` , as one is saved in our store, and the other one has no persistence. Basically, the `StorageBackend` is ephemeral, only there to hide how we store the file. Once we implement persistence, instead of storing the temporary files inside a dict, we will rely on the OS filesystem to retrieve the files. Right now, it seems like a lot of duplication as we need to store the ephemeral state inside `StorageBackend`